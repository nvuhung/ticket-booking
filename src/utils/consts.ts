export type SnackBarType = 'danger' | 'warning' | 'success';
export const MaxTicket = 6;
export const Currency = 'Ä‘';
