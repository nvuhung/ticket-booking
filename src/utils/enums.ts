export enum TicketTypes {
  Standard,
  Vip,
  Deluxe
}

export enum TicketStatuses {
  Available,
  Booked,
  Selecting
}
