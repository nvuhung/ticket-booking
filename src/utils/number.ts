import { Currency } from './consts';

export const formatCurrency = (value: number = 0, hasSuffix: boolean = true) =>
  value.toString().replace(/(\d)(?=(\d{3})+(?!\d))/g, '$1.') + (hasSuffix ? Currency : '');
