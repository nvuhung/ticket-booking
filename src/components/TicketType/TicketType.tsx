import React from 'react';
import styled from 'styled-components';

import { ITicketTypes } from '../../models/ticket-type';
import { theme } from '../../styles/theme';
import { formatCurrency } from '../../utils/number';

export interface IProps {
  types: ITicketTypes[];
}

const TicketType: React.FC<IProps> = ({ types }) => (
  <Container>
    {types.map(item => (
      <TypeItem className="type-item" key={item.id}>
        <TypeColor color={theme.colors.types[item.id]} />
        <Text className="type-text">
          {item.description} - {formatCurrency(item.price)}
        </Text>
      </TypeItem>
    ))}
  </Container>
);

const Container = styled.div`
  margin-left: 20px;
`;

const TypeItem = styled.div`
  display: flex;
  align-items: center;
  margin-top: 10px;
`;

const TypeColor = styled.div<{ color: string }>`
  width: 18px;
  height: 18px;
  border-radius: 50%;
  border: solid 1px ${props => props.color};
  margin-right: 10px;
`;

const Text = styled.span`
  color: ${theme.colors.white};
`;

export default React.memo(TicketType);
