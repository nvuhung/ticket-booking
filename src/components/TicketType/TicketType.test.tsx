import React from 'react';
import 'jest-styled-components';

import TicketType, { IProps } from './TicketType';
import { mount, shallow } from '../../test/enzyme';
import { seats } from '../../test/mocks';
import { formatCurrency } from '../../utils/number';

describe('TicketType', () => {
  let props: IProps = (null as any) as IProps;
  let mountedComp: any;

  const wrapper = (bShallow: boolean) => {
    if (!mountedComp) {
      mountedComp = bShallow
        ? shallow(<TicketType {...props} />)
        : mount(<TicketType {...props} />);
    }
    return mountedComp;
  };

  beforeEach(() => {
    props = {
      types: seats.types
    };
  });

  it('renders without crashing', () => {
    const comp = wrapper(false);
    expect(comp).toMatchSnapshot();
  });

  describe('when movie is null', () => {
    beforeEach(() => {
      props.types = [];
      mountedComp = undefined;
    });

    it('should not render', () => {
      const comp = wrapper(false);
      expect(comp.find('div > div').length).toEqual(0);
    });
  });

  describe('when types has value', () => {
    beforeEach(() => {
      props.types = seats.types;
      mountedComp = undefined;
    });

    it('should render contents', () => {
      const comp = wrapper(false);
      const elts = comp.find('div.type-item');
      expect(elts.length).toEqual(props.types.length);
    });

    it('should match data', () => {
      const comp = wrapper(false);
      const first = comp
        .find('div.type-item')
        .first()
        .find('.type-text')
        .first();
      expect(first.text()).toEqual(
        `${seats.types[0].description} - ${formatCurrency(seats.types[0].price)}`
      );
    });
  });
});
