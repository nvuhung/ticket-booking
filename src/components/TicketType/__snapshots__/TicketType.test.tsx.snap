// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TicketType renders without crashing 1`] = `
.c0 {
  margin-left: 20px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-top: 10px;
}

.c2 {
  width: 18px;
  height: 18px;
  border-radius: 50%;
  border: solid 1px #fff;
  margin-right: 10px;
}

.c4 {
  width: 18px;
  height: 18px;
  border-radius: 50%;
  border: solid 1px #385846;
  margin-right: 10px;
}

.c5 {
  width: 18px;
  height: 18px;
  border-radius: 50%;
  border: solid 1px #384b5d;
  margin-right: 10px;
}

.c3 {
  color: #fff;
}

<Memo(TicketType)
  types={
    Array [
      Object {
        "description": "Standard",
        "id": 0,
        "price": 60000,
      },
      Object {
        "description": "VIP",
        "id": 1,
        "price": 90000,
      },
      Object {
        "description": "Deluxe",
        "id": 2,
        "price": 100000,
      },
    ]
  }
>
  <styled.div>
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-bdVaJa",
            "isStatic": false,
            "lastClassName": "c0",
            "rules": Array [
              "
  margin-left: 20px;
",
            ],
          },
          "displayName": "styled.div",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-bdVaJa",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
    >
      <div
        className="c0"
      >
        <styled.div
          className="type-item"
          key="0"
        >
          <StyledComponent
            className="type-item"
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-bwzfXH",
                  "isStatic": false,
                  "lastClassName": "c1",
                  "rules": Array [
                    "
  display: flex;
  align-items: center;
  margin-top: 10px;
",
                  ],
                },
                "displayName": "styled.div",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-bwzfXH",
                "target": "div",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
          >
            <div
              className="c1 type-item"
            >
              <styled.div
                color="#fff"
              >
                <StyledComponent
                  color="#fff"
                  forwardedComponent={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "attrs": Array [],
                      "componentStyle": ComponentStyle {
                        "componentId": "sc-htpNat",
                        "isStatic": false,
                        "lastClassName": "c5",
                        "rules": Array [
                          "
  width: 18px;
  height: 18px;
  border-radius: 50%;
  border: solid 1px ",
                          [Function],
                          ";
  margin-right: 10px;
",
                        ],
                      },
                      "displayName": "styled.div",
                      "foldedComponentIds": Array [],
                      "render": [Function],
                      "styledComponentId": "sc-htpNat",
                      "target": "div",
                      "toString": [Function],
                      "warnTooManyClasses": [Function],
                      "withComponent": [Function],
                    }
                  }
                  forwardedRef={null}
                >
                  <div
                    className="c2"
                    color="#fff"
                  />
                </StyledComponent>
              </styled.div>
              <styled.span
                className="type-text"
              >
                <StyledComponent
                  className="type-text"
                  forwardedComponent={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "attrs": Array [],
                      "componentStyle": ComponentStyle {
                        "componentId": "sc-bxivhb",
                        "isStatic": false,
                        "lastClassName": "c3",
                        "rules": Array [
                          "
  color: ",
                          "#fff",
                          ";
",
                        ],
                      },
                      "displayName": "styled.span",
                      "foldedComponentIds": Array [],
                      "render": [Function],
                      "styledComponentId": "sc-bxivhb",
                      "target": "span",
                      "toString": [Function],
                      "warnTooManyClasses": [Function],
                      "withComponent": [Function],
                    }
                  }
                  forwardedRef={null}
                >
                  <span
                    className="c3 type-text"
                  >
                    Standard
                     - 
                    60.000đ
                  </span>
                </StyledComponent>
              </styled.span>
            </div>
          </StyledComponent>
        </styled.div>
        <styled.div
          className="type-item"
          key="1"
        >
          <StyledComponent
            className="type-item"
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-bwzfXH",
                  "isStatic": false,
                  "lastClassName": "c1",
                  "rules": Array [
                    "
  display: flex;
  align-items: center;
  margin-top: 10px;
",
                  ],
                },
                "displayName": "styled.div",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-bwzfXH",
                "target": "div",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
          >
            <div
              className="c1 type-item"
            >
              <styled.div
                color="#385846"
              >
                <StyledComponent
                  color="#385846"
                  forwardedComponent={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "attrs": Array [],
                      "componentStyle": ComponentStyle {
                        "componentId": "sc-htpNat",
                        "isStatic": false,
                        "lastClassName": "c5",
                        "rules": Array [
                          "
  width: 18px;
  height: 18px;
  border-radius: 50%;
  border: solid 1px ",
                          [Function],
                          ";
  margin-right: 10px;
",
                        ],
                      },
                      "displayName": "styled.div",
                      "foldedComponentIds": Array [],
                      "render": [Function],
                      "styledComponentId": "sc-htpNat",
                      "target": "div",
                      "toString": [Function],
                      "warnTooManyClasses": [Function],
                      "withComponent": [Function],
                    }
                  }
                  forwardedRef={null}
                >
                  <div
                    className="c4"
                    color="#385846"
                  />
                </StyledComponent>
              </styled.div>
              <styled.span
                className="type-text"
              >
                <StyledComponent
                  className="type-text"
                  forwardedComponent={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "attrs": Array [],
                      "componentStyle": ComponentStyle {
                        "componentId": "sc-bxivhb",
                        "isStatic": false,
                        "lastClassName": "c3",
                        "rules": Array [
                          "
  color: ",
                          "#fff",
                          ";
",
                        ],
                      },
                      "displayName": "styled.span",
                      "foldedComponentIds": Array [],
                      "render": [Function],
                      "styledComponentId": "sc-bxivhb",
                      "target": "span",
                      "toString": [Function],
                      "warnTooManyClasses": [Function],
                      "withComponent": [Function],
                    }
                  }
                  forwardedRef={null}
                >
                  <span
                    className="c3 type-text"
                  >
                    VIP
                     - 
                    90.000đ
                  </span>
                </StyledComponent>
              </styled.span>
            </div>
          </StyledComponent>
        </styled.div>
        <styled.div
          className="type-item"
          key="2"
        >
          <StyledComponent
            className="type-item"
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-bwzfXH",
                  "isStatic": false,
                  "lastClassName": "c1",
                  "rules": Array [
                    "
  display: flex;
  align-items: center;
  margin-top: 10px;
",
                  ],
                },
                "displayName": "styled.div",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-bwzfXH",
                "target": "div",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
          >
            <div
              className="c1 type-item"
            >
              <styled.div
                color="#384b5d"
              >
                <StyledComponent
                  color="#384b5d"
                  forwardedComponent={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "attrs": Array [],
                      "componentStyle": ComponentStyle {
                        "componentId": "sc-htpNat",
                        "isStatic": false,
                        "lastClassName": "c5",
                        "rules": Array [
                          "
  width: 18px;
  height: 18px;
  border-radius: 50%;
  border: solid 1px ",
                          [Function],
                          ";
  margin-right: 10px;
",
                        ],
                      },
                      "displayName": "styled.div",
                      "foldedComponentIds": Array [],
                      "render": [Function],
                      "styledComponentId": "sc-htpNat",
                      "target": "div",
                      "toString": [Function],
                      "warnTooManyClasses": [Function],
                      "withComponent": [Function],
                    }
                  }
                  forwardedRef={null}
                >
                  <div
                    className="c5"
                    color="#384b5d"
                  />
                </StyledComponent>
              </styled.div>
              <styled.span
                className="type-text"
              >
                <StyledComponent
                  className="type-text"
                  forwardedComponent={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "attrs": Array [],
                      "componentStyle": ComponentStyle {
                        "componentId": "sc-bxivhb",
                        "isStatic": false,
                        "lastClassName": "c3",
                        "rules": Array [
                          "
  color: ",
                          "#fff",
                          ";
",
                        ],
                      },
                      "displayName": "styled.span",
                      "foldedComponentIds": Array [],
                      "render": [Function],
                      "styledComponentId": "sc-bxivhb",
                      "target": "span",
                      "toString": [Function],
                      "warnTooManyClasses": [Function],
                      "withComponent": [Function],
                    }
                  }
                  forwardedRef={null}
                >
                  <span
                    className="c3 type-text"
                  >
                    Deluxe
                     - 
                    100.000đ
                  </span>
                </StyledComponent>
              </styled.span>
            </div>
          </StyledComponent>
        </styled.div>
      </div>
    </StyledComponent>
  </styled.div>
</Memo(TicketType)>
`;
